### ISSUE TITLE
Fails to build with OpenSSL 1.1.0

### ISSUE THREAD

**Issue by ppisar** [dsully/perl-crypt-openssl-x509#53]
After upgrading OpenSSL to 1.1.0b the build fails like this:

```
gcc -c  -I/usr/include/openssl -I/usr/local/include/ssl -I/usr/local/ssl/include -D_REENTRANT -D_GNU_SOURCE -O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches -m32 -march=i686 -mtune=atom -fasynchronous-unwind-tables -fwrapv -fno-strict-aliasing -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -g   -DVERSION=\"1.807\" -DXS_VERSION=\"1.807\" -fPIC "-I/usr/lib/perl5/CORE"   X509.c
X509.xs: In function '_decode_netscape':
X509.xs:217:5: error: unknown type name 'NETSCAPE_X509'
     NETSCAPE_X509 nx;
     ^~~~~~~~~~~~~
```

**Comment by sebastianas** [comment]
https://github.com/dsully/perl-crypt-openssl-x509/pulls

can you get this tested please?

**Comment by ppisar** [comment]
It emmits these fatal warnings:

gcc -c  -I/usr/include/openssl -I/usr/local/include/ssl -I/usr/local/ssl/include -D_REENTRANT -D_GNU_SOURCE -O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches -m64 -mtune=generic -fwrapv -fno-strict-aliasing -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -O2 -g -Wall -Werror   -DVERSION=\"1.807\" -DXS_VERSION=\"1.807\" -fPIC "-I/usr/lib64/perl5/CORE"   X509.c
X509.xs: In function ‘XS_Crypt__OpenSSL__X509_accessor’:
X509.xs:462:43: error: passing argument 1 of ‘OPENSSL_sk_num’ from incompatible pointer type [-Werror=incompatible-pointer-types]
     for (j = 0; j < sk_OPENSSL_STRING_num(emlst); j++) {
                                           ^~~~~
In file included from /usr/include/openssl/crypto.h:28:0,
                 from /usr/include/openssl/bio.h:20,
                 from /usr/include/openssl/asn1.h:16,
                 from X509.xs:5:
/usr/include/openssl/stack.h:23:5: note: expected ‘const OPENSSL_STACK * {aka const struct stack_st *}’ but argument is of type ‘struct stack_st_OPENSSL_STRING *’
 int OPENSSL_sk_num(const OPENSSL_STACK *);
     ^~~~~~~~~~~~~~
X509.xs:463:53: error: passing argument 1 of ‘OPENSSL_sk_value’ from incompatible pointer type [-Werror=incompatible-pointer-types]
       BIO_printf(bio, "%s", sk_OPENSSL_STRING_value(emlst, j));
                                                     ^~~~~
In file included from /usr/include/openssl/crypto.h:28:0,
                 from /usr/include/openssl/bio.h:20,
                 from /usr/include/openssl/asn1.h:16,
                 from X509.xs:5:
/usr/include/openssl/stack.h:24:7: note: expected ‘const OPENSSL_STACK * {aka const struct stack_st *}’ but argument is of type ‘struct stack_st_OPENSSL_STRING *’
 void *OPENSSL_sk_value(const OPENSSL_STACK *, int);
       ^~~~~~~~~~~~~~~~
X509.xs:463:25: error: format ‘%s’ expects argument of type ‘char *’, but argument 3 has type ‘void *’ [-Werror=format=]
       BIO_printf(bio, "%s", sk_OPENSSL_STRING_value(emlst, j));
                         ^
cc1: all warnings being treated as errors

If I remove the -Werror, it builds and all tests pass. So I think it's almost
good.

-- Petr

**Comment by ppisar** [comment]
The <https://github.com/dsully/perl-crypt-openssl-x509/pull/55> fixes the warning. Tested with OpenSSL 1.1.0b and 1.0.2j.

**Comment by dur-randir** [comment]
Hi @dsully, can you release a version to CPAN with this fix? Both Debian 9 and Archlinux are out with openssl 1.1.0, so this module can't be built using cpan on them, you have to install from git.

**Comment by dsully** [comment]
Not yet - there's another build problem:

```x86_64-linux-gnu-gcc -c  -I/usr/local/opt/openssl/include -I/usr/include/openssl -I/usr/local/include/ssl -I/usr/local/ssl/include -D_REENTRANT -D_GNU_SOURCE -DDEBIAN -fwrapv -fno-strict-aliasing -pipe -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -O2 -g -Wall -Werror   -DVERSION=\"1.807\" -DXS_VERSION=\"1.807\" -fPIC "-I/usr/lib/x86_64-linux-gnu/perl/5.22/CORE"   X509.c
X509.xs:81:13: error: conflicting types for ‘X509_get0_signature’
 static void X509_get0_signature(const_ossl11 ASN1_BIT_STRING **psig, const_ossl11 X509_ALGOR **palg,
             ^~~~~~~~~~~~~~~~~~~
In file included from /usr/include/openssl/pem.h:70:0,
                 from X509.xs:11:
/usr/include/openssl/x509.h:837:6: note: previous declaration of ‘X509_get0_signature’ was here
 void X509_get0_signature(ASN1_BIT_STRING **psig, X509_ALGOR **palg,
      ^~~~~~~~~~~~~~~~~~~
Makefile:347: recipe for target 'X509.o' failed
make: *** [X509.o] Error 1```

**Comment by cypa** [comment]
+1
got same error on python 2.7 on openbsd 6.3

**Comment by toddr** [comment]
This issue looks to be resolved now?

**Comment by ppisar** [comment]
I think all the necessary fixes were released in 1.808 version.
